src/compute/src/sink/copy_to_s3_oneshot.rs

```diff
@@ -67,68 +81,215 @@ where
+                            }
+                        }
+                    }
+                }
+                AsyncEvent::Progress(frontier) => {
+                    if PartialOrder::less_equal(&up_to, &frontier) {
+                        match uploader.flush().await {
+                            Ok(()) => {}
+                            Err(e) => {
+                                send_response(CopyToResponse::Error(e.to_string()));
+                                return;
+                            }
+                        }
+                        // We are done, send the final count.
+                        send_response(CopyToResponse::RowCount(row_count));
+                        return;
+                    }
                 }
             }
         }
     });
 }
 
+/// Required state to upload batches to S3
+struct CopyToS3Uploader {
+    /// The output description.
+    desc: RelationDesc,
+    /// Params to format the data.
+    format: CopyFormatParams<'static>,
+    /// The index of the next file.
+    file_index: usize,
+    /// The prefix for the file names.
+    file_name_prefix: String,
+    /// The bucket prefix batches should be uploaded to.
+    path_prefix: String,
+    /// A buffer of encoded data.
+    buf: Vec<u8>,
+    /// The desired batch size. When the buffer exceeds this value a put operation will be
+    /// performed.
+    max_file_size: u64,
+    /// The aws sdk config.
+    sdk_config: SdkConfig,
+}
+
+impl CopyToS3Uploader {
+    fn extract_s3_bucket_path(prefix: &str) -> (String, String) {
+        let uri = Uri::from_str(prefix).expect("valid s3 url");
+        let bucket = uri.host().expect("s3 bucket");
+        let path = uri.path().trim_start_matches('/').trim_end_matches('/');
+        (bucket.to_string(), path.to_string())
+    }
+
+    /// Flushes the internal buffer to S3 by uploading a file. If there is no pending data to
+    /// be uploaded this is a no-op.
+    async fn flush(&mut self) -> Result<(), anyhow::Error> {
+        if !self.buf.is_empty() {
+            self.file_index += 1;
+            let (bucket, path_prefix) = Self::extract_s3_bucket_path(&self.path_prefix);
+            // TODO: remove hard-coded file extension for .csv
+            let file_path = format!(
+                "{}/{}-{:04}.csv",
+                path_prefix, self.file_name_prefix, self.file_index
+            );
+            info!("starting upload at bucket: {}, file {}", bucket, file_path);
+            let mut uploader = S3MultiPartUploader::try_new(
+                &self.sdk_config,
+                bucket,
+                file_path,
+                S3MultiPartUploaderConfig {
+                    part_size_limit: ByteSize::mib(10).as_u64(),
+                    file_size_limit: self.max_file_size,
+                },
+            )
+            .await?;
+
+            uploader.add_chunk(Bytes::from(self.buf.clone())).await?;
+            uploader.finish().await?;
```

Comment:
Currently, I am not particularly leveraging the multi part uploader. This is still waiting to upload only after buffering an entire file's worth of data when it instead can buffer only a part's worth of data and upload the part. I will refactor this later.